### ISSUE TITLE
False positive on `no_std` for `unconditional_recursion`  lint

### ISSUE THREAD

**Issue by rossmacarthur** [rust-lang/rust#80332]
I tried this code:

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=71ed3d2cb8eeedd2386b80a4eb9e57da)

```rust
#![no_std]

trait Abs {
    fn abs(self) -> Self;
}

impl Abs for f64 {
    fn abs(self) -> f64 {
        core::primitive::f64::abs(self)
    }
}
```

I expected to see it compile with no warnings.

Instead, the following lint was triggered

```
warning: function cannot return without recursing
 --> src/lib.rs:8:5
  |
8 |     fn abs(self) -> f64 {
  |     ^^^^^^^^^^^^^^^^^^^ cannot return without recursing
9 |         core::primitive::f64::abs(self)
  |         ------------------------------- recursive call site
  |
  = note: `#[warn(unconditional_recursion)]` on by default
  = help: a `loop` may express intention better if this is on purpose
```

This is a false positive because

```rust
assert_eq!(Abs::abs(-1.0), 1.0);
```

works completely fine and does not keep recursing.

### Meta

At the time of opening this issue this lint was incorrectly triggered on `no_std` for latest stable, beta, and nightly, and at least as far back as v1.20.

**Comment by SNCPlay42** [comment]
This is in fact correct: there is no `abs` method for `f64` in `core`, only in `std`, so in a `#![no_std]` context `f64::abs` refers to the function you have written. If you look at the assembly for the code given on [Godbolt](https://rust.godbolt.org/z/rW58Pb) you can see that the function is indeed calling itself.

**Comment by rossmacarthur** [comment]
@SNCPlay42 Ah I see, woops. I think I got confused. Could you explain to me why `cargo test` works here

```rust
#![no_std]

#[test]
fn example() {
    let x: f64 = -1.0;
    assert_eq!(x.abs(), 1.0);
}
```

But `cargo build` fails here

```rust
#![no_std]

fn example() {
    let x: f64 = -1.0;
    assert_eq!(x.abs(), 1.0);
}
```

Does cargo force linking to `std` when using `cfg(test)`?
